### DISCUSSION TITLE
4D Timeline animation with custom start date

### DISCUSSION THREAD

**Discussion by oussamabonnor1** [iTwin/itwinjs-core#3491] [Q&A] [answered]
Is it possible to start the timeline as close as possible to today's date ?

The result should be like this: 

![image](https://user-images.githubusercontent.com/17766221/162201724-35867452-74af-420f-b6d8-db28a69b5577.png)

Is this code snippet going in the right direction by the way ?

```
const vp = IModelApp.viewManager.selectedView;
const was = vp?.view.scheduleScript?.modelTimelines[0];
```

Also, how can I get access to [TimeLineComponent](https://www.itwinjs.org/v2/reference/ui-components/timeline/timelinecomponent/)?

Finally, please consider adding a sample example on how to manipulate schedules within iModels.

**Comment by shenshengyi** [comment]
Here is my imodel schedules video url.
https://www.bilibili.com/video/BV1BS4y1Z7pB?spm_id_from=333.999.0.0
https://www.bilibili.com/video/BV1qq4y1w7Ns?spm_id_from=333.999.0.0

**Reply by oussamabonnor1** [reply]
Thanks for the links! I already have a working 4D schedule integrated. I was wondering how to make it start from any point I choose (not from the beginning).

**Comment by pmconne** [comment, marked as answer]
We have been putting off making a sample demonstrating schedule scripts because we first want to remove the need to store the script in the iModel (vs defining and applying it on the frontend). That work is in progress now.

@bsteinbk @NancyMcCallB for questions about TimelineComponent.

**Reply by oussamabonnor1** [reply]
Alright great, thanks for the answer. Meanwhile, I'd appreciate a way to manipulate that start date or schedule even if I end up using a script saved in the imodel.

My goal is to start the timeline from today's date to see the current expected state of the BIM.

**Reply by pmconne** [reply]
I assume your timeline component is being produced by the [DefaultViewOverlay](https://www.itwinjs.org/reference/appui-react/contentview/defaultviewoverlay/) supplied by [IModelViewportControl](https://www.itwinjs.org/learning/ui/appui-react/imodelviewportcontrol/). Are you using that control directly or indirectly e.g. via iTwin Viewer?
I am not a UI person, but I do not see an obvious way to programmatically access the timeline component via the viewport control.
@NancyMcCallB and/or @bsteinbk?

If there's not a better solution you could consider assigning to [Viewport.timePoint](https://www.itwinjs.org/reference/core-common/displaystyles/displaystylesettings/timepoint/) when first opening the view, to point it to the current date and time (or any other time). You'd probably want to confirm the viewport's display style has a schedule script and that the time point falls within the script's duration.

**Reply by oussamabonnor1** [reply]
The [Viewport.timePoint](https://www.itwinjs.org/reference/core-common/displaystyles/displaystylesettings/timepoint/) solution is perfect! Thank you so much!
